### DISCUSSION TITLE
After upgrading my application with spring-shell-starter to 3.1.3 not able to pass command line arguments

### DISCUSSION THREAD

**Discussion by suryaachandra** [spring-projects/spring-shell#842] [Q&A] [answered]
My application was running with ``` spring-shell-starter: 2.0.1.RELEASE ``` , I have upgraded to ``` spring-shell-starter: 3.1.3 ``` today 

when my application was running with spring-shell-starter: 2.0.1.RELEASE I tried to bring up the application by passing command line arguments as  below, the application used to come up successfully accepting the arguments 

```  java -jar .\target\devbox-shell-0.0.1-SNAPSHOT.jar user=xyx password=123 ```

once the application comes up, I used to type help command and try all the commands 

But after I upgraded with spring-shell-starter: 3.1.3 and when I ran the application using the same 

``` java -jar .\target\devbox-shell-0.0.1-SNAPSHOT.jar user=xyx password=123 ```

I am getting the below error, it is not accepting as arguments instead looking as a command 

```
[  :  :  ][main]/[2023-08-03 18:48:27,654][devbox-shell][ INFO][c.h.s.g.DevBoxShellApplication][] Started DevBoxShellApplication in 3.864 seconds (process running for 4.817)
No command found for 'user=xyx password=123'
[  :  :  ][main]/[2023-08-03 18:48:27,717][devbox-shell][ INFO][o.s.b.a.l.ConditionEvaluationReportLogger][]

Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
[  :  :  ][main]/[2023-08-03 18:48:27,781][devbox-shell][ERROR][o.s.b.SpringApplication][] Application run failed
org.springframework.shell.CommandNotFound: No command found for 'user=xyx password=123'
        at org.springframework.shell.Shell.evaluate(Shell.java:207)
        at org.springframework.shell.Shell.run(Shell.java:159)
        at org.springframework.shell.jline.NonInteractiveShellRunner.run(NonInteractiveShellRunner.java:146)
        at org.springframework.shell.DefaultShellApplicationRunner.run(DefaultShellApplicationRunner.java:65)
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:762)
Wrapped by: java.lang.IllegalStateException: Failed to execute ApplicationRunner
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:765)
        at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:752)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:319)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1295)

```

Please help me how to pass arguments ifor the latest shell

**Comment by jvalkeal** [comment]
Older shell version is radically different what comes for command line parsing. It's difficult to say what goes wrong as you didn't show how your command(s) are defined. I can comment more if you show how what your application does.

**Reply by suryaachandra** [reply]
We have created an application that connects to APM and fetches the data and writes to CSV files. we used to pass apm.read_token=xyz as argument while running the application.

**Case 1 (Before Upgrading): Application with Spring Boot 2.7, Java 11, and Spring Shell-[2.0.1.RELEASE]**

we tried to bring up the application for this case as below 

`java -jar demo.jar apm.read_token=xyz`

this used to work for this case is taking as an argument and was set in the application context applications comes up and a shell terminal was created and next when we run the ap.topology command to call the API and token is fetched from the application context and passed while running the command and used to execute without any issue. 

Below is the code for our command defined

```
    @ShellMethod(value = "Topology", key = "apm.topology")
    public String topology(String serviceName, String env) throws IOException {

        return (serviceName.equals("*")) ? getTopology(env).toString()
                : getTopology(serviceName, env).toString();
    }
```


```
  private ApmModel.ServiceTopology getTopology(String service, String env) throws IOException {
        return ApmApis.getTopology(service, getToken("apm.read_token"), buildRequest(env));
    }
```

	
```
 public String getToken(String tokenName){
        var args= Arrays.stream( applicationContext.getBean(ApplicationArguments.class).getSourceArgs()).filter(v->v.startsWith(tokenName)).collect(Collectors.toList());
        if(args.size()>0){
            return Arrays.stream(args.get(0).split("=")).collect(Collectors.toList()).get(1);
        }
        return "Empty";
    }
```


**Case 2 (After Upgrading) : Application with Spring Boot 3.1.2, Java 17, and Spring Shell-3.1.3**

we have upgraded our application with the latest Spring boot, java, and spring shell and we tried to run the application using the same command we are getting command not found issue, with this latest upgrade its not taking apm.read_token=xyz as argument but is expecting some command to be available.

`java -jar demo.jar apm.read_token=xyz`

we are getting this below error 

`
No command found for 'apm.read_token=xyz'
[  :  :  ][main]/[2023-08-09 10:08:07,418][servicekernal-shell][ INFO][o.s.b.a.l.ConditionEvaluationReportLogger][]

Error starting ApplicationContext. To display the condition evaluation report re-run your application with 'debug' enabled.
[  :  :  ][main]/[2023-08-09 10:08:07,446][servicekernal-shell][ERROR][o.s.b.SpringApplication][] Application run failed
org.springframework.shell.CommandNotFound: No command found for 'apm.read_token=xyz'
        at org.springframework.shell.Shell.evaluate(Shell.java:207)
        at org.springframework.shell.Shell.run(Shell.java:159)
        at org.springframework.shell.jline.NonInteractiveShellRunner.run(NonInteractiveShellRunner.java:146)
        at org.springframework.shell.DefaultShellApplicationRunner.run(DefaultShellApplicationRunner.java:65)
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:762)
Wrapped by: java.lang.IllegalStateException: Failed to execute ApplicationRunner
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:765)
        at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:752)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:319)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306)
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1295)

`

We have tried different possibilities like -- and - to pass arguments but everytime it is expecting to be command while bringing up the application.

please help to fix this issue.

**Comment by donqing** [comment]
I have this problem too, so how fix it?

**Reply by suryaachandra** [reply, marked as answer]
We got the issued fixed in our application by setting the below property in applications.properties file ( convert to yaml if you are using yaml). Now it is working as before.

`
spring.shell.noninteractive.enabled=false
`
